qbraid_qir/qasm3/visitor.py

```diff
@@ -657,7 +692,101 @@ def _visit_generic_gate_operation(self, operation: QuantumGate) -> None:
             else:
                 self._visit_basic_gate_operation(operation, inverse_value)
 
-    def _visit_classical_operation(self, statement: ClassicalDeclaration) -> None:
+    def _validate_variable_assignment_value(self, variable: Variable, value: Any) -> None:
+        """Validate the assignment of a value to a variable.
+
+        Args:
+            variable (Variable): The variable to assign to.
+            value (Any): The value to assign.
+
+        Raises:
+            Qasm3ConversionError: If the value is not of the correct type.
+        """
+        # check 1 - type match
+
+        base_type = variable.base_type
+        qasm_type = base_type.__class__
+        try:
+            type_to_match = VARIABLE_TYPE_MAP[qasm_type]
+        except KeyError:
+            raise Qasm3ConversionError(f"Invalid type {qasm_type} for variable {variable.name}")
+
+        if not isinstance(value, type_to_match):
+            raise Qasm3ConversionError(
+                f"Invalid assignment of type {type(value)} to variable {variable.name}"
+            )
+
+        # check 2 - range match , if bits mentioned in base size
+        if type_to_match == int:
+            base_size = variable.base_size
+            left, right = 0, 0
+            if qasm_type == Qasm3IntType:
+                left, right = -1 * (2 ** (base_size - 1)), 2 ** (base_size - 1) - 1
+            else:
+                # would be uint only then we correctly get this
+                left, right = 0, 2**base_size - 1
+            if value < left or value > right:
+                raise Qasm3ConversionError(
+                    f"Value {value} out of limits for variable {variable.name} with base size {base_size}"
+                )
+
+        elif type_to_match == float:
+            base_size = variable.base_size
+            left, right = 0, 0
+            # Reference : https://openqasm.com/language/types.html#floating-point-numbers and IEEE 754 standard
+            if base_size == 32:
+                left, right = -(1.70141183 * (10**38)), (1.70141183 * (10**38))
+            else:
+                left, right = -(10**308), (10**308)
```

Comment:
I looked at the openqasm link but it's still not clear to me where the 1.7 and 38 / 308 numbers are coming from? Could you explain a bit?

Reply:
Yep! These are the floating point value limits according to the IEEE 754 specification. The reference which I used for these is the [wiki page](https://en.wikipedia.org/wiki/Single-precision_floating-point_format#:~:text=A%20signed%2032%2Dbit%20integer,%E2%89%88%203.4028235%20%C3%97%201038.)

It defines the limits for the `float[32]` and `float[64]` values. Since openqasm uses IEEE 754 convention, I defined the limits accordingly.